oxidation/crates/parsec_core_fs/src/storage/chunk_storage.rs

```diff
@@ -0,0 +1,404 @@
+
+#[derive(Insertable, AsChangeset)]
+#[diesel(table_name = chunks)]
+struct NewChunk<'a> {
+    pub chunk_id: &'a [u8],
+    pub size: i32,
+    pub offline: bool,
+    pub accessed_on: Option<f32>,
+    pub data: &'a [u8],
+}
+
+pub(crate) trait ChunkStorageTrait {
+    fn conn(&self) -> &Mutex<SqliteConn>;
+    fn local_symkey(&self) -> &SecretKey;
+
+    fn create_db(&self) -> FSResult<()> {
+        let conn = &mut *self.conn().lock().expect("Mutex is poisoned");
+        sql_query(
+            "CREATE TABLE IF NOT EXISTS chunks (
+                chunk_id BLOB PRIMARY KEY NOT NULL, -- UUID
+                size INTEGER NOT NULL,
+                offline BOOLEAN NOT NULL,
+                accessed_on REAL, -- Timestamp
+                data BLOB NOT NULL
+            );",
+        )
+        .execute(conn)
+        .map_err(|_| FSError::CreateTable("chunks"))?;
+        Ok(())
+    }
+
+    #[cfg(test)]
+    fn drop_db(&self) -> FSResult<()> {
+        let conn = &mut *self.conn().lock().expect("Mutex is poisoned");
+        sql_query("DROP TABLE IF EXISTS chunks;")
+            .execute(conn)
+            .map_err(|_| FSError::DropTable("chunks"))?;
+        Ok(())
+    }
+
+    // Size and chunks
+
+    fn get_nb_blocks(&self) -> FSResult<i64> {
+        let conn = &mut *self.conn().lock().expect("Mutex is poisoned");
+        chunks::table
+            .select(count_star())
+            .first(conn)
+            .map_err(|_| FSError::QueryTable("chunks: get_nb_blocks"))
+    }
+
+    fn get_total_size(&self) -> FSResult<i64> {
+        let conn = &mut *self.conn().lock().expect("Mutex is poisoned");
+        chunks::table
+            .select(coalesce_total_size())
+            .first(conn)
+            .map_err(|_| FSError::QueryTable("chunks: get_total_size"))
+    }
+
+    fn is_chunk(&self, chunk_id: ChunkID) -> FSResult<bool> {
+        let conn = &mut *self.conn().lock().expect("Mutex is poisoned");
+        chunks::table
+            .select(count_star())
+            .filter(chunks::chunk_id.eq((*chunk_id).as_ref()))
+            .first::<i64>(conn)
+            .map_err(|_| FSError::QueryTable("chunks: is_chunk"))
+            .map(|res| res > 0)
+    }
+
+    fn get_local_chunk_ids(&self, chunk_ids: &[ChunkID]) -> FSResult<Vec<ChunkID>> {
+        let bytes_id_list = chunk_ids
+            .iter()
+            .map(|chunk_id| (**chunk_id).as_ref())
+            .collect::<Vec<_>>();
+
+        let mut res = Vec::with_capacity(chunk_ids.len());
+
+        let conn = &mut *self.conn().lock().expect("Mutex is poisoned");
+        for bytes_id_list_chunk in bytes_id_list.chunks(SQLITE_MAX_VARIABLE_NUMBER) {
+            res.append(
+                &mut chunks::table
+                    .select(chunks::chunk_id)
+                    .filter(chunks::chunk_id.eq_any(bytes_id_list_chunk))
+                    .load::<Vec<u8>>(conn)
+                    .map_err(|_| FSError::QueryTable("chunks: get_local_chunk_ids"))?
```

Comment:
could be useful to save the error returned by `load::<Vec<u8>>(conn)` for debug purpose